### PR TITLE
feature: top_k selective_log_softmax

### PR DESCRIPTION
**Pull request by LeonEricsson**
Generalize `selective_log_softmax` to support gathering multiple (`top_k`) log-probabilities per position.

`selective_log_softmax` now accepts `index` with shape `(..., K)` **or** `(...)`:

* `(..., K)` gathers **K** log-probabilities per position (new multi-index path).
* `(...)` existing single-token shorthand; internally treated as `K=1` via unsqueeze/squeeze so the output shape matches the input `index` shape.

This change is fully backward compatible and strictly extends the function’s functionality by enabling retrieval of multiple (e.g., top-k) log-probabilities when needed.

---

For completeness, benchmarks across increasing `top_k` values, where `K=1` is equivalent to the legacy implementation.

<details>
<summary>benchmark code</summary>

```
import time

import torch
import torch.nn.functional as F


def selective_log_softmax(logits, index) -> torch.Tensor:
    squeeze = index.ndim == logits.ndim - 1
    if squeeze:
        index = index.unsqueeze(-1)

    if logits.dtype in [torch.float32, torch.float64]:
        selected_logits = torch.gather(logits, dim=-1, index=index)
        # loop to reduce peak mem consumption
        logsumexp_values = torch.stack([torch.logsumexp(lg, dim=-1) for lg in logits])
        per_token_logps = selected_logits - logsumexp_values.unsqueeze(-1)  # log_softmax(x_i) = x_i - logsumexp(x)
    else:
        # logsumexp approach is unstable with bfloat16, fall back to slightly less efficient approach
        per_token_logps = []
        for row_logits, row_labels in zip(logits, index, strict=True):  # loop to reduce peak mem consumption
            row_logps = F.log_softmax(row_logits, dim=-1)
            row_per_token_logps = row_logps.gather(dim=-1, index=row_labels)
            per_token_logps.append(row_per_token_logps)
        per_token_logps = torch.stack(per_token_logps)

    if squeeze:
        per_token_logps = per_token_logps.squeeze(-1)
    return per_token_logps


def new_wrapper_topk(logits, input_ids, logits_to_keep, k=8):
    logits = logits[:, -logits_to_keep:]
    index = torch.topk(logits, k=k, dim=-1).indices
    return selective_log_softmax(logits, index)


def measure_memory_and_time(func, logits, input_ids, logits_to_keep):
    torch.cuda.reset_peak_memory_stats()
    start_time = time.time()
    result = func(logits, input_ids, logits_to_keep)
    end_time = time.time()
    mem_peak = torch.cuda.max_memory_allocated()
    return result, end_time - start_time, mem_peak


def fmt_mem(b):
    return f"{b / 1024**2:.1f} MB"


def main():
    device = "cuda"
    vocab_size = 150000
    seq_len = 512
    batch_size = 8
    logits_to_keep = 128
    warmup_iters = 3
    bench_iters = 10

    for dtype in [torch.float32, torch.bfloat16]:
        print(f"\n{'=' * 70}")
        print(f"dtype={dtype}, batch={batch_size}, seq_len={seq_len}, vocab={vocab_size}")
        print(f"{'=' * 70}")

        logits = torch.randn(batch_size, seq_len, vocab_size, dtype=dtype, device=device)
        input_ids = torch.randint(0, vocab_size, (batch_size, seq_len), device=device)

        for k in [1, 2, 4, 8, 16, 20]:
            if k == 1:
                # K=1: use the no-K-dim shorthand (index shape (...))
                func = lambda l, i, ltk: selective_log_softmax(l[:, -ltk:], i[:, -ltk:])
            else:
                _k = k  # capture loop variable
                func = lambda l, i, ltk, _k=_k: selective_log_softmax(
                    l[:, -ltk:], torch.topk(l[:, -ltk:], k=_k, dim=-1).indices
                )
            name = f"K={k}"

            # warmup
            for _ in range(warmup_iters):
                func(logits, input_ids, logits_to_keep)
            torch.cuda.synchronize()

            times, mems = [], []
            for _ in range(bench_iters):
                torch.cuda.empty_cache()
                _, t, m = measure_memory_and_time(func, logits, input_ids, logits_to_keep)
                torch.cuda.synchronize()
                times.append(t)
                mems.append(m)

            avg_time = sum(times) / len(times)
            avg_mem = sum(mems) / len(mems)
            print(f"  {name:20s}  time={avg_time * 1000:8.2f} ms  peak_mem={fmt_mem(avg_mem)}")


if __name__ == "__main__":
    main()
```

</details>

```
======================================================================
dtype=torch.float32, batch=8, seq_len=512, vocab=150000
======================================================================
  K=1                   time=    1.29 ms  peak_mem=2418.0 MB
  K=2                   time=    3.48 ms  peak_mem=2935.2 MB
  K=4                   time=    3.46 ms  peak_mem=2935.2 MB
  K=8                   time=    3.47 ms  peak_mem=2935.2 MB
  K=16                  time=    3.47 ms  peak_mem=2935.3 MB
  K=20                  time=    3.47 ms  peak_mem=2935.4 MB

======================================================================
dtype=torch.bfloat16, batch=8, seq_len=512, vocab=150000
======================================================================
  K=1                   time=    2.06 ms  peak_mem=1245.3 MB
  K=2                   time=    3.25 ms  peak_mem=1470.1 MB
  K=4                   time=    3.29 ms  peak_mem=1470.1 MB
  K=8                   time=    3.30 ms  peak_mem=1470.2 MB
  K=16                  time=    3.28 ms  peak_mem=1470.3 MB
  K=20                  time=    3.29 ms  peak_mem=1470.3 MB
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/utils.py [hunk 1]
```diff
@@ -635,7 +635,7 @@ def flush_right(mask: torch.Tensor, *tensors: torch.Tensor) -> torch.Tensor | tu
     return flushed_mask, *flushed_tensors
 
 
-def selective_log_softmax(logits, index) -> torch.Tensor:
+def selective_log_softmax2(logits, index) -> torch.Tensor:
```
Comment (qgallouedec) [inline trl/trainer/utils.py]: this is meant to be removed, right?

Comment (qgallouedec) [reply trl/trainer/utils.py]: I mean, the whole old function

Comment (LeonEricsson) [reply trl/trainer/utils.py]: woops

**Comment by qgallouedec** [comment]
I'm fine with merging this PR, but to get some more context, what the motivation behind it? this new feature seems not to be used anywhere

**Comment by qgallouedec** [comment]
also, can you run 

```
pip install hf-doc-builder
make precommit
```

**Comment by LeonEricsson** [comment]
fyi: I need this functionality for an upcoming PR, but since its self-contained and easy to review i figured it could be its own PR

File: trl/trainer/utils.py [hunk 2]
```diff
@@ -641,32 +641,41 @@ def selective_log_softmax(logits, index) -> torch.Tensor:
 
     This function is equivalent to the following naive implementation:
     ```python
-    logps = torch.gather(logits.log_softmax(-1), dim=-1, index=index.unsqueeze(-1)).squeeze(-1)
+    logps = torch.gather(logits.log_softmax(-1), dim=-1, index=index)
```
Comment (qgallouedec) [inline trl/trainer/utils.py]:
```suggestion
    logps = torch.gather(logits.log_softmax(-1), dim=-1, index=index.unsqueeze(-1)).squeeze(-1)  # for index with shape (...)
    logps = torch.gather(logits.log_softmax(-1), dim=-1, index=index)  # for index with shape (..., K)
```
